WalletWasabi.Fluent/ViewModels/Wallets/Home/Tiles/PrivacyRing/PrivacyBarViewModel.cs

```diff
@@ -89,69 +83,101 @@ private void CreateSegments(IExtendedList<PrivacyBarItemViewModel> list, IEnumer
 
 	private IEnumerable<PrivacyBarItemViewModel> CreateCoinSegments(IEnumerable<Pocket> pockets, int coinCount)
 	{
-		var total = pockets.Sum(x => Math.Abs(x.Amount.ToDecimal(NBitcoin.MoneyUnit.BTC)));
-		var start = 0.0m;
+		var totalAmount = pockets.Sum(x => Math.Abs(x.Amount.ToDecimal(NBitcoin.MoneyUnit.BTC)));
+		var usableWidth = (decimal)Width - (coinCount - 1) * _gapBetweenSegments;
+
+		// Calculate the width of the segments.
+		var rawSegments = pockets
+			.Where(x => x.Coins.Any())
+			.SelectMany(pocket => pocket.Coins)
+			.Select(coin =>
+			{
+				var pocket = pockets.First(pocket => pocket.Coins.Contains(coin));
+				var amount = coin.Amount.ToDecimal(NBitcoin.MoneyUnit.BTC);
+				var width = Math.Abs(usableWidth * amount / totalAmount);
 
-		var usableWidth = Width - (coinCount * 2);
+				return (OwnerPocket: pocket, Coin: coin, Width: width);
+			}).ToArray();
 
-		var usablePockets =
-				pockets.Where(x => x.Coins.Any())
-					   .OrderByDescending(x => x.Coins.First().HdPubKey.AnonymitySet)
-					   .ToList();
+		var segmentsToEnlarge = rawSegments.Where(x => x.Width < 2).ToArray();
 
-		foreach (var pocket in usablePockets)
+		// Artificially enlarge segments smaller than 2 px in order to make them visible.
+		// Meanwhile decrease those segments that are larger than 2 px in order to fit all in the bar.
+		if (segmentsToEnlarge.Any())
 		{
-			var pocketCoins = pocket.Coins.OrderByDescending(x => x.Amount).ToList();
+			var enlargeBy = 1m;
+			var segmentsToReduce = rawSegments.Except(segmentsToEnlarge);
+			var reduceBy = segmentsToEnlarge.Length * enlargeBy / segmentsToReduce.Count();
 
-			foreach (var coin in pocketCoins)
+			rawSegments = rawSegments.Select(x =>
 			{
-				var margin = 2;
-				var amount = coin.Amount.ToDecimal(NBitcoin.MoneyUnit.BTC);
-				var width = Math.Abs((decimal)usableWidth * amount / total);
+				var finalWidth = x.Width < 2 ? x.Width + enlargeBy : x.Width - reduceBy;
+				return (OwnerPocket: x.OwnerPocket, Coin: x.Coin, Width: finalWidth);
+			}).ToArray();
+		}
 
-				// Artificially enlarge segments smaller than 2 px in order to make them visible.
-				if (width < 2)
-				{
-					width++;
-					margin--;
-				}
+		// Order the coins as they will be shown in the bar.
+		rawSegments = rawSegments
+			.OrderByDescending(x => x.OwnerPocket.Coins.First().HdPubKey.AnonymitySet)
+			.ThenByDescending(x => x.Width)
+			.ToArray();
 
-				yield return new PrivacyBarItemViewModel(this, coin, (double)start, (double)width);
+		var start = 0.0m;
+		foreach (var tup in rawSegments)
+		{
+			var (_, coin, width) = tup;
```

Comment:
nit: you can deconstruct directly in the foreach:

```suggestion
		foreach (var (_, coin, width) in rawSegments)
		{

```